File: cocos/3d/ui/assembler/mask/mask-assembler.ts
/****************************************************************************
 Copyright (c) 2017-2018 Xiamen Yaji Software Co., Ltd.

 http://www.cocos.com

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated engine source code (the "Software"), a limited,
 worldwide, royalty-free, non-assignable, revocable and non-exclusive license
 to use Cocos Creator solely to develop games on your target platforms. You shall
 not use Cocos Creator software for developing other software or tools that's
 used for developing games. You are not granted to publish, distribute,
 sublicense, and/or sell copies of Cocos Creator.

 The software or tools in this License Agreement are licensed, not sold.
 Xiamen Yaji Software Co., Ltd. reserves all rights not expressly granted to you.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
 ****************************************************************************/

import { Mat4, Vec3 } from '../../../../core/value-types';
import { vec3 } from '../../../../core/vmath';
import { IRenderData, RenderData } from '../../../../renderer/ui/renderData';
import { UI } from '../../../../renderer/ui/ui';
import { Node } from '../../../../scene-graph/node';
import { MaskComponent } from '../../components';
import { IAssembler, IAssemblerManager } from '../assembler';
import { StencilManager } from './stencil-manager';

const _stencilManager = StencilManager.sharedManager!;
const _temp_vec3 = new Vec3();
const _worldMatrix = new Mat4();

export const maskAssembler: IAssembler = {
    createData (mask: MaskComponent) {
        const renderData = mask.requestRenderData();
        renderData!.dataLength = 4;
        renderData!.vertexCount = 4;
        renderData!.indiceCount = 6;
        return renderData as RenderData;
    },

    updateRenderData (mask: MaskComponent){
        const renderData = mask.renderData;
        if (renderData) {
            if (renderData.vertDirty) {
                if (this.updateVerts) {
                    this.updateVerts(mask);
                }
            }
        }
    },

    updateVerts (mask: MaskComponent) {
        const renderData: RenderData | null = mask.renderData;
        if (!renderData) {
            return;
        }

        const node: Node = mask.node;
        const datas: IRenderData[] = renderData.datas;
        const cw = node.width;
        const ch = node.height;
        const appx = node.anchorX * cw;
        const appy = node.anchorY * ch;
        let l = 0;
        let b = 0;
        let r = 0;
        let t = 0;
        // if (sprite.trim) {
        l = -appx;
        b = -appy;
        r = cw - appx;
        t = ch - appy;
        datas[0].x = l;
        datas[0].y = b;
        datas[3].x = r;
        datas[3].y = t;

        vec3.set(_temp_vec3, l, b, 0);
        const rect = renderData.rect;
        mask.node.getWorldMatrix(_worldMatrix);
        vec3.transformMat4(_temp_vec3, _temp_vec3, _worldMatrix);
        rect.origin = _temp_vec3;
        rect.width = r - l;
        rect.height = t - b;

        renderData.vertDirty = false;
    },

    fillBuffers (mask: MaskComponent, renderer: UI) {
        _stencilManager.pushMask(mask);
        // commit last data
        renderer.autoMergeBatches();

        _stencilManager.clear();
        mask.clearGraphics!.updateAssembler(renderer);

        _stencilManager.enterLevel();
        mask.graphics!.updateAssembler(renderer);

        _stencilManager.enableMask();
    },
};

export const maskEndAssembler: IAssembler = {
    fillBuffers (mask: MaskComponent, ui: UI) {
        _stencilManager.exitMask();
    },
};

const StartAssembler: IAssemblerManager = {
    getAssembler () {
        return maskAssembler;
    },
};

const PostAssembler: IAssemblerManager = {
    getAssembler () {
        return maskEndAssembler;
    },
};

MaskComponent.Assembler = StartAssembler;
MaskComponent.PostAssembler = PostAssembler;

File: native/cocos/renderer/scene/assembler/MaskAssembler.cpp
/****************************************************************************
 Copyright (c) 2018 Xiamen Yaji Software Co., Ltd.

 http://www.cocos2d-x.org
 
 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:
 
 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.
 
 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
 ****************************************************************************/

#include "MaskAssembler.hpp"
#include "../ModelBatcher.hpp"
#include "../StencilManager.hpp"
#include "../../Macro.h"

RENDERER_BEGIN

MaskAssembler::MaskAssembler()
{
}

MaskAssembler::~MaskAssembler()
{
    CC_SAFE_RELEASE(_renderSubHandle);
    CC_SAFE_RELEASE(_clearSubHandle);
}

void MaskAssembler::setRenderSubHandle(Assembler* renderSubHandle)
{
    if (_renderSubHandle == renderSubHandle) return;
    CC_SAFE_RELEASE(_renderSubHandle);
    _renderSubHandle = renderSubHandle;
    CC_SAFE_RETAIN(_renderSubHandle);
}

void MaskAssembler::setClearSubHandle(Assembler* clearSubHandle)
{
    if (_clearSubHandle == clearSubHandle) return;
    CC_SAFE_RELEASE(_clearSubHandle);
    _clearSubHandle = clearSubHandle;
    CC_SAFE_RETAIN(_clearSubHandle);
}

void MaskAssembler::handle(NodeProxy *node, ModelBatcher* batcher, Scene* scene)
{
    batcher->flush();
    batcher->flushIA();

    StencilManager* instance = StencilManager::getInstance();
    instance->pushMask(_inverted);
    instance->clear();
    batcher->commit(node, _clearSubHandle);
    batcher->flush();
    instance->enterLevel();

    if (_imageStencil)
    {
        batcher->commit(node, this);
    }
    else if (_renderSubHandle)
    {
        _renderSubHandle->handle(node, batcher, scene);
    }

    batcher->flush();
    instance->enableMask();
}

void MaskAssembler::postHandle(NodeProxy *node, ModelBatcher *batcher, Scene *scene)
{
    batcher->flush();
    batcher->flushIA();
    StencilManager::getInstance()->exitMask();
}

RENDERER_END
